### PR TITLE
UCT/UCM: Fix loadlable module inter-library dependencies

### LINKED ISSUE #3319: Issues with dynamically loaded uct modules on master
Hi, 

@TomAugspurger and I first noticed the following error on master when we tried to use listen-connect-accept API (which depends on rdmacm uct module for functioning)
```
$ UCX_TLS=all python3 benchmarks/recv-into-client.py -o numpy --n-iter=100
ucp_listener.c:263  UCX  ERROR none of the available transports can listen for connections on 0.0.0.0:6452
failed to listen (Destination is unreachable)
Failed to failed to start listener
```
This python program depends on the the ucx-py library with python bindings for UCX: `build/lib.linux-x86_64-3.7/ucp_py/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so` which has the following dependencies

```
$ ldd build/lib.linux-x86_64-3.7/ucp_py/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so 
        linux-vdso.so.1 =>  (0x00007ffde108c000)
        libucp.so.0 => $UCX_HOME/lib/libucp.so.0 (0x00002b7b9ef23000)
        libucs.so.0 => $UCX_HOME/lib/libucs.so.0 (0x00002b7b9f1e2000)
        libcudart.so.9.2 => $CUDA_HOME/lib64/libcudart.so.9.2 (0x00002b7b9f436000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00002b7b9f6a0000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00002b7b9f8bd000)
        libuct.so.0 => $UCX_HOME/lib/libuct.so.0 (0x00002b7b9fc87000)
        libnuma.so.1 => /usr/lib/x86_64-linux-gnu/libnuma.so.1 (0x00002b7b9feb2000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00002b7ba00bd000)
        libucm.so.0 => $UCX_HOME/lib/libucm.so.0 (0x00002b7ba03c6000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00002b7ba05e3000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00002b7ba07eb000)
        /lib64/ld-linux-x86-64.so.2 (0x00002b7b9ea76000)
```

When specific UCX_TLS are requested, the following warnings are omitted in addition to the error: 
```
$ UCX_TLS=rc,rdmacm,sm,tcp,cuda_copy,cuda_ipc,gdr_copy python3 benchmarks/recv-into-client.py -o numpy
[1551821475.563082]     ucp_context.c:600  UCX  WARN  transport 'rc' is not available
[1551821475.563111]     ucp_context.c:600  UCX  WARN  transport 'rdmacm' is not available
[1551821475.563117]     ucp_context.c:600  UCX  WARN  transport 'cuda_copy' is not available
[1551821475.563123]     ucp_context.c:600  UCX  WARN  transport 'cuda_ipc' is not available
[1551821475.563128]     ucp_context.c:600  UCX  WARN  transport 'gdr_copy' is not available
```

The installation has all the modules in lib/ucx folder
```
$ ls $UCX_HOME/build/lib/ucx
libucm_cuda.a             libuct_cuda_gdrcopy.so.0.0.0  libuct_ib_cm.so.0.0.0  libuct_rdmacm.a
libucm_cuda.la            libuct_cuda.la                libuct_ib.la           libuct_rdmacm.la
libucm_cuda.so            libuct_cuda.so                libuct_ib.so           libuct_rdmacm.so
libucm_cuda.so.0          libuct_cuda.so.0              libuct_ib.so.0         libuct_rdmacm.so.0
libucm_cuda.so.0.0.0      libuct_cuda.so.0.0.0          libuct_ib.so.0.0.0     libuct_rdmacm.so.0.0.0
libuct_cuda.a             libuct_ib.a                   libuct_knem.a          libucx_perftest_cuda.a
libuct_cuda_gdrcopy.a     libuct_ib_cm.a                libuct_knem.la         libucx_perftest_cuda.la
libuct_cuda_gdrcopy.la    libuct_ib_cm.la               libuct_knem.so         libucx_perftest_cuda.so
libuct_cuda_gdrcopy.so    libuct_ib_cm.so               libuct_knem.so.0       libucx_perftest_cuda.so.0
libuct_cuda_gdrcopy.so.0  libuct_ib_cm.so.0             libuct_knem.so.0.0.0   libucx_perftest_cuda.so.0.0.0
```
I'm unable to reproduce this issue when I manually build `ucp_client_server.c` in `test/examples` using:
```
gcc test/examples/ucp_client_server.c -o $PWD/examples-bin/ucp_client_server.out -Itest/examples -I$UCX_HOME/include -L $UCX_HOME/lib -lucm -lucs -lucp -luct -Wl,-rpath="-L$UCX_HOME/lib" -L$CUDA_HOME/lib64 -lcuda -lcudart -I$CUDA_HOME/include
```
One of the main differences between how ucp_client_server example is complied and how `build/lib.linux-x86_64-3.7/ucp_py/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so` is compiled is that the latter passes  `-shared -fPIC` to gcc (which is a Cython requirement if I'm not wrong)

I contacted @bureddy  about this and he asked me to run with `UCX_LOG_LEVEL=DEBUG` and it like some symbols are undefined:
```
[1551822388.524287]         module.c:60   UCX  DEBUG ucs library path: $UCX_HOME/lib/libucs.so.0
[1551822388.524306]         module.c:185  UCX  DEBUG loading modules for uct
[1551822388.527325]         module.c:168  UCX  DEBUG could not load '$UCX_HOME/lib/ucx/libuct_cuda.so.0': $UCX_HOME/lib/ucx/libuct_cuda.so.0: undefined symbol: void_class
[1551822388.529049]         module.c:168  UCX  DEBUG could not load '$UCX_HOME/lib/ucx/libuct_cuda.so.0': $UCX_HOME/lib/ucx/libuct_cuda.so.0: undefined symbol: void_class
[1551822388.531174]         module.c:168  UCX  DEBUG could not load '$UCX_HOME/lib/ucx/libuct_ib.so.0': $UCX_HOME/lib/ucx/libuct_ib.so.0: undefined symbol: ucs_mpool_put
[1551822388.532300]         module.c:168  UCX  DEBUG could not load '$UCX_HOME/lib/ucx/libuct_ib.so.0': $UCX_HOME/lib/ucx/libuct_ib.so.0: undefined symbol: ucs_mpool_put
[1551822388.534370]         module.c:168  UCX  DEBUG could not load '$UCX_HOME/lib/ucx/libuct_rdmacm.so.0': $UCX_HOME/lib/ucx/libuct_rdmacm.so.0: undefined symbol: void_class
[1551822388.535523]         module.c:168  UCX  DEBUG could not load '$UCX_HOME/lib/ucx/libuct_rdmacm.so.0': $UCX_HOME/lib/ucx/libuct_rdmacm.so.0: undefined symbol: void_class
[1551822388.537088]         module.c:168  UCX  DEBUG could not load '$UCX_HOME/lib/ucx/libuct_knem.so.0': $UCX_HOME/lib/ucx/libuct_knem.so.0: undefined symbol: void_class
```

Does any of this help? Do you have thoughts on what could be the problem and how we can get around it?

### PR DESCRIPTION
**Pull request by yosefe**
Fixes #3319

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9230/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9231/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9233/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9250/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9254/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9261/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9271/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9273/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9279/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9287/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9289/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9290/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9300/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@hoopoepg @brminich can you pls take a look?

File: test/apps/test_ucp_dlopen.c [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    if (ptr2 == MAP_FAILED) {
+        printf("mmmap() failed: %m\n");
+        return -1;
+    }
+
+    /* load ucp */
+    printf("opening '%s'\n", filename);
+    handle = dlopen(filename, RTLD_NOW | RTLD_LOCAL);//TODO no GLOBAL
+    if (handle == NULL) {
+        fprintf(stderr, "failed to open %s: %m\n", filename);
```
Comment (brminich) [inline test/apps/test_ucp_dlopen.c]: ` test_ucp_init` prints errors to stdout. would be better to use one approach

File: contrib/test_jenkins.sh [hunk 2]
```diff
@@ -832,12 +832,28 @@ test_ucs_load() {
...
+	../contrib/configure-release --prefix=$ucx_inst
+	$MAKEP clean
+	$MAKEP
+
+	# Make sure UCP library, when opened with dlopen(), loads CMA module
+	if find -name libuct_cma.so.0
+	then
+		echo "==== Running UCP library loading test ===="
+		./test/apps/test_ucp_dlopen # just to save output to log
+		./test/apps/test_ucp_dlopen | grep 'cma/cma'
```
Comment (brminich) [inline contrib/test_jenkins.sh]: what is the aim of this grep?

File: src/ucs/sys/module.c [hunk 3]
```diff
@@ -115,6 +115,39 @@ static void ucs_module_loader_init_paths()
...
+    ret = dladdr(addr, &dl_info);
+    if (ret == 0) {
+        ucs_module_debug("dladdr(%p) [%s] failed: %s", addr, symbol, dlerror());
+        return NULL;
+    }
+
+    /* return the symbol only if it was found in the requested library, and not,
+     * for example, in one of its dependencies.
+     */
+    if (strcmp(dl_info.dli_fname, module_path)) {
```
Comment (hoopoepg) [inline src/ucs/sys/module.c]: minor: does it guarantee that dl_info.dli_fname has absolute path? even in case if it points to symbolic link?

Comment (yosefe) [reply src/ucs/sys/module.c]: good point... i will check if we can compare something more strict than the path

File: contrib/test_jenkins.sh (hunk 2 above)
Comment (yosefe) [reply contrib/test_jenkins.sh]: it should check that cma is loaded, by check that UCP has cma dev on cma transport (hence cma/cma)

File: src/ucs/sys/module.c (hunk 3 above)
Comment (hoopoepg) [reply src/ucs/sys/module.c]: readlink + fstat -> compare inodes

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9320/ for details (Mellanox internal link).

**Comment by Akshay-Venkatesh** [comment]
@yosefe We see the issue that was reported with dynamic module loading on master even with this fix. Wanted to check if this PR work in-part or was it supposed to fix the ucxpy problem we reported? 

Context: Coming back to this issue because we were trying to use ucx-1.5.x with our `tcpcm` (https://github.com/Akshay-Venkatesh/ucx/tree/topic/tcpcm) work (generated on top of master) but seems like master and 1.5.x have diverged a fair bit.

**Pull request merged**